### ISSUE TITLE
`ComplexMethod` rule question

### ISSUE THREAD

**Issue by lwasylkowski** [arturbosch/detekt#680]
I don't quite understand why this method:

```kotlin
private fun getNetworkType(): String =
        when (telephonyManager.networkType) {
            TelephonyManager.NETWORK_TYPE_1xRTT -> "1xRTT"
            TelephonyManager.NETWORK_TYPE_CDMA -> "CDMA"
            TelephonyManager.NETWORK_TYPE_EDGE -> "EDGE"
            TelephonyManager.NETWORK_TYPE_EHRPD -> "eHRPD"
            TelephonyManager.NETWORK_TYPE_EVDO_0 -> "EVDO rev. 0"
            TelephonyManager.NETWORK_TYPE_EVDO_A -> "EVDO rev. A"
            TelephonyManager.NETWORK_TYPE_EVDO_B -> "EVDO rev. B"
            TelephonyManager.NETWORK_TYPE_GPRS -> "GPRS"
            TelephonyManager.NETWORK_TYPE_HSDPA -> "HSDPA"
            TelephonyManager.NETWORK_TYPE_HSPA -> "HSPA"
            TelephonyManager.NETWORK_TYPE_HSPAP -> "HSPA+"
            TelephonyManager.NETWORK_TYPE_HSUPA -> "HSUPA"
            TelephonyManager.NETWORK_TYPE_IDEN -> "iDen"
            TelephonyManager.NETWORK_TYPE_LTE -> "LTE"
            TelephonyManager.NETWORK_TYPE_UMTS -> "UMTS"
            TelephonyManager.NETWORK_TYPE_UNKNOWN -> "Unknown"
            else -> "Unknown"
        }
```

gets reported as `ComplexMethod - 18/10`. There is only one `when`, and I understand there are many clauses, but I thought this method should have complexity of like `2`, because there's only one `when` and no nested flow controls

**Comment by schalkms** [comment]
Because there are 18 case branches in your example. This is the behavior I would expect. I hope that answers your question.

**Comment by arturbosch** [comment]
This is a behavior we discussed earlier in an issue. It is really tricky in the case of a when statement.
I think we should maybe only count the branch in a when case if it has a block statement. Or we could add a property which will count a when statement as 2 ... What do you think?

**Comment by schalkms** [comment]
I like the way how it’s done at the moment but I am open for new better ideas.

**Comment by lwasylkowski** [comment]
I just realised that this check is a regular McCabe cyclomatic complexity, for whatever reason I was under impression that this is some arbitrarily chosen method. In this case I'm against changing it, but maybe I'd change the name to `CyclomaticComplexity` or `McCabeComplexity` for clarity. 

The issue can be closed as far as I'm concerned, unless you want to discuss it further. Sorry for making the fuss!

**Comment by schalkms** [comment]
Thanks for raising that question!
If **ComplexMethod** should implement the McCabe complexity then this behavior is right. However, it is criticized that such a simple method with a switch case reports such a high complexity.

**Comment by arturbosch** [comment]
Hey, jap mccabe is used here. ComplexMethod is the name which is often used in code smell papers, so I adapted it. I do not know if all developers are familiar with software metrics even if mccabe is like the standard metric together with .

**Comment by vanniktech** [comment]
PMD also hast those McCabe as rule names.

**Comment by scottkennedy** [comment]
I have mixed feelings about this.  I like keeping code simple, but I currently have a function that's nearly identical to the one here, and I just don't see a way to make it any less "complex".  Because it's not complex.

**Comment by arturbosch** [comment]
There could be a property which says "do not count a path in when-expression if it is not a block statement".
What do you think?

**Comment by arturbosch** [comment]
There could be a property which says "do not count a path in when-expression if it is not a block statement".
What do you think?

**Comment by scottkennedy** [comment]
I'd be in favor of that.  Anything to be able to keep this option enabled, without catching simple `when` blocks.

**Comment by schalkms** [comment]
I think the rule could ignore a method iff there is a single when statement in the body. This should be enabled via a config option.

**Comment by arturbosch** [comment]
That's also a good idea!
